### ISSUE TITLE
[HOLD for payment 2024-07-24] [$250] [Uneven Splits] [Polish] Android  - Split amount is not reset when tapping Reset button when keyboard is up

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#41763]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.71-0
**Reproducible in staging?:** y
**Reproducible in production?:** new feature
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
1. Launch New Expensify app.
2. Go to FAB > Split expense.
3. Enter amount and select a few users.
4. Tap on split amount input.
5. Manually enter the split amount.
6. With keyboard up, tap on Reset button.

## Expected Result:
The split amount will be reset (mweb behavior when keyboard is up).

## Actual Result:
The keyboard is dismissed and the split amount is not reset.

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/bc3628b3-3d50-4a64-aff8-c68eef43c839

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010af563a3cc74c7c2</li>
        <li>Upwork Job ID: 1787911203897004032</li>
        <li>Last Price Increase: 2024-05-28</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by m-natarajan** [comment]
@isabelastisser FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by m-natarajan** [comment]
We think this bug might be related to #vip-split

**Comment by rlinoz** [comment]
This is related to the new uneven split feature https://github.com/Expensify/App/pull/40386

I don't think this needs to be a blocker, as the reset feature still works, but curious for @youssef-lr's thoughts.

Also, are you taking this issues or should I make it external?

**Comment by youssef-lr** [comment]
Thanks @amyevans, let's open this up for contributors as I haven't been able to figure it out yet and don't have time to do so.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Can't press the reset button when the keyboard is visible.

### What is the root cause of that problem?
It's the default behavior of a list view that when we press, the keyboard will close and the button won't receive the click/press. To customize that, we have [keyboardShouldPersistTaps](https://reactnative.dev/docs/scrollview#keyboardshouldpersisttaps) prop. We can set it to `always` so the view can receive click/press while the keyboard is shown.

The list used in the money request confirmation page is BaseOptionsSelector which uses BaseOptionsList. SectionList in BaseOptionsList already set it to always
https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/components/OptionsList/BaseOptionsList.tsx#L256

but we have a nested scroll view here that doesn't set the `keyboardShouldPersistTaps`.
https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/components/OptionsSelector/BaseOptionsSelector.js#L632-L640

### What changes do you think we should make in order to solve the problem?

Set both ScrollView `keyboardShouldPersistTaps` to 'always'. This will reset the split amount, however, the focused input amount will only be updated after it's blurred because of `formatAmountOnBlur` prop. To fix it, we can manually dismiss the keyboard using `Keyboard.dismiss` in here.
https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/components/MoneyRequestConfirmationList.tsx#L528

https://github.com/Expensify/App/assets/50919443/af5c7c14-aaa5-4179-b0a1-925bed411bb7

**Comment by youssef-lr** [comment]
@bernhardoj is that the recording of the solution? if yes then I think we'll still have to tap twice on Reset?

**Comment by kaushiktd** [comment]
### Please re-state the problem that we are trying to solve in this issue.
Android - Split - Split amount is not reset when tapping Reset button when keyboard is up

### What is the root cause of that problem?
The issue occurred because the resetSplitShares function was not resetting the split shares as expected. This function is responsible for updating the split shares of a transaction.  it calls another function called setSplitShares, which actually performs the update by setting new amounts and currencies for each split share.

### What changes do you think we should make in order to solve the problem?

After looking into the matter more closely, we discovered that the issue stemmed from the resetSplitShares function failing to dismiss the keyboard after resetting the split shares. By integrating the Keyboard.dismiss() statement directly within the function, we managed to resolve this inconsistency. This adjustment ensured that the keyboard was consistently dismissed after the split shares were reset, effectively addressing the issue at hand
add after this line
https://github.com/Expensify/App/blob/807c945ef8edabe86cc62f16c866aa754db8750b/src/libs/actions/IOU.ts#L6461
 `Keyboard.dismiss();`

video:-https://drive.google.com/file/d/1-TxdUKxdgpY2oIPUvc_B7Gj7vn7QdBbJ/view?usp=sharing

**Comment by youssef-lr** [comment]
>Nope, only once.

Ah okay. So it seems like there is a slight delay due to having to dismiss the keyboard first. Anyway we can eliminate it @bernhardoj or there's really no way around it?

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by bernhardoj** [comment]
@youssef-lr eliminate which part? The keyboard dismiss?

**Comment by youssef-lr** [comment]
The slight delay from when we press Reset and the shares are actually reset

**Comment by kaushiktd** [comment]
@youssef-lr  I found the issue could be resolved without altering ScrollView components. Including Keyboard.dismiss() within resetSplitShares function effectively addressed the problem.

**Comment by bernhardoj** [comment]
Hmm, looks like the delay is because we dismissed the keyboard. Maybe the animation delays the onyx update. I think we need to either rework the `formatOnBlur` so the amount is updated when we reset it.

Or we can let the onyx complete first then dismiss the keyboard.
`IOU.resetSplitShares(transaction).then(Keyboard.dismiss)`

https://github.com/Expensify/App/assets/50919443/ace4cf3e-9172-4c73-a6f3-6b76ff722fe7

**Comment by isabelastisser** [comment]
@eVoloshchak, please provide an update. Thanks!

**Comment by eVoloshchak** [comment]
According to the issue description
> Expected Result: The split amount will be reset (mweb behavior when keyboard is up).

Is the expected behavior is for the keyboard to stay open when the 'Reset' button is pressed? Seems like the desired behavior, but that's not the current behavior on mWeb

@youssef-lr, could you confirm what is the expected behavior here please?

**Comment by youssef-lr** [comment]
>Is the expected behavior is for the keyboard to stay open when the 'Reset' button is pressed? Seems like the desired behavior, but that's not the current behavior on mWeb

This would be great if possible

**Comment by eVoloshchak** [comment]
@bernhardoj, @kaushiktd, do you think [this](https://github.com/Expensify/App/issues/41763#issuecomment-2111432505) is achievable?

**Comment by kaushiktd** [comment]
@youssef-lr 
By simply adding the Keyboard.dismiss() statement within the resetSplitShares function, we effectively addressed the problem.

Currently, the keyboard behavior is such that it hides when the 'Reset' button is pressed. 

Regarding the expected behavior, if the desired outcome is for the keyboard to remain open when the 'Reset' button is pressed, this deviates from the current solution where the keyboard is dismissed. If keeping the keyboard open is preferred, we can explore alternative solutions to achieve this while ensuring the split amounts are properly reset."

**Comment by kaushiktd** [comment]
@eVoloshchak  i update [my Proposa](https://github.com/Expensify/App/issues/41763#issuecomment-2102708992)l

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by youssef-lr** [comment]
bump @eVoloshchak

**Comment by isabelastisser** [comment]
Hey @eVoloshchak, I DM'd you for visibility.

**Comment by eVoloshchak** [comment]
@kaushiktd, thank you for the updated proposal! The behavior on the [video](https://drive.google.com/file/d/1qBw0WyJtHdWR95SnQEXA8N2R7LgpkT2l/view?usp=sharing) matches the expected behavior, but the implementation can be improved

> To solve the problem, we need to introduce a flag resetClicked, which is set to true when the reset button is clicked. This flag will indicate that the input field's value should be reset.

**_Why_** do we need to do that?
 Let's dig deeper and figure out the actual root cause for this, introducing an additional flag does look like a workaround. 
Is there a wrong `newAmount` passed to [setNewAmount](https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/components/MoneyRequestAmountInput.tsx#L131) function?

**Comment by kaushiktd** [comment]
> **_Why_** do we need to do that?

The problem occurs because the input field remains focused, implying it's within the setNewAmount function. To signal a reset button click, we must set a resetClicked flag. This flag enables setNewAmount to detect the reset event and update the input field's value accordingly. Interestingly, in the web version, the reset button click doesn't maintain focus on the input field, ensuring proper detection of the reset event and correct value updating.

there is no problem in newAmount passed to setNewAmount function

**Comment by youssef-lr** [comment]
To give some context, the amount is not updated if the input is focused. I think we should find a way to bypass this if the amount change is triggered from the reset button. The reason I did this is to not have the amount auto formatted while the user was typing it.

**Comment by kaushiktd** [comment]
@youssef-lr  
My code ensures that the auto-formatting functionality remains unaffected while allowing the input field to reset the amount when the reset button is clicked. Additionally, as you can see in the [video](https://drive.google.com/file/d/1qBw0WyJtHdWR95SnQEXA8N2R7LgpkT2l/view?usp=sharing), the cursor remains focused on the input field after the reset

**Comment by youssef-lr** [comment]
That's looking good, can we have the cursor move to the end when the amount is reset?

**Comment by youssef-lr** [comment]
@kaushiktd also do you have a test branch?

**Comment by kaushiktd** [comment]
> That's looking good, can we have the cursor move to the end when the amount is reset?

Let me dig up something!

And what do you mean by test branch?

**Comment by kaushiktd** [comment]
@youssef-lr  i update [my proposal](https://github.com/Expensify/App/issues/41763#issuecomment-2102708992)

**Comment by isabelastisser** [comment]
@eVoloshchak @youssef-lr can you please provide an update? Thanks!

**Comment by youssef-lr** [comment]
I think the proposal from @kaushiktd makes sense to me, please tag me once the PR is up

**Comment by kaushiktd** [comment]
Your Expensify account email: prakash.techdoodles@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01e7f7b7a75799e9db

**Comment by kaushiktd** [comment]
My previous comment is for bot because I'm getting error in receiving offer for last couple of issue. So just tried to update the details.

**Comment by kaushiktd** [comment]
@youssef-lr A PR for another issue is already in progress and ir is not merged yet so how can I create new PR without affecting that PR? I am not able to see fork option because that one is still in progress.

**Comment by isabelastisser** [comment]
Bump @eVoloshchak and @youssef-lr to see @kaushiktd comments above. Thanks!

**Comment by youssef-lr** [comment]
@kaushiktd what other PR? can you link it here please?

**Comment by kaushiktd** [comment]
@youssef-lr It was for another issue and I've already submitted PR for this one and tagged you there as well

**Comment by youssef-lr** [comment]
>A PR for another issue is already in progress and PR is not merged yet so how can I create new PR without affecting that PR?

Ah I think I see, is this about our guidelines to only have one PR in progress?

**Comment by kaushiktd** [comment]
@youssef-lr Yes but I got it sorted from slack.

**Comment by isabelastisser** [comment]
@eVoloshchak @youssef-lr @kaushiktd, what's next here? Is the PR on hold?

**Comment by kaushiktd** [comment]
@isabelastisser PR is ready. Waiting for @eVoloshchak review.

**Comment by isabelastisser** [comment]
Thanks, @kaushiktd! @eVoloshchak, please review the PR. I will DM you for visibility. Thanks!

**Comment by kaushiktd** [comment]
@isabelastisser working on new PR

**Comment by isabelastisser** [comment]
I will be OOO tomorrow and next week, so I am reassigning this until I return on July 29. Thanks, @alexpensify 

Status: 

- @eVoloshchak needs to complete the BZ list.
- Payment is due on July 23 via NewDot.

**Comment by alexpensify** [comment]
Noted, I'll prepare next week for the payment date.

**Comment by alexpensify** [comment]
Not overdue, the payment date is on the 24th.

**Comment by alexpensify** [comment]
Correct me if I'm wrong, but it looks like there is another PR in the works associated with this GH: https://github.com/Expensify/App/pull/45333. I'm going to hold off on completing the payment action until all PRs are in production. 

@kaushiktd - please apply here: 
https://www.upwork.com/jobs/~016c93f917170cbc76 

Thanks!

**Comment by kaushiktd** [comment]
@alexpensify Is it possible for you to send me an invitation to that job?

**Comment by alexpensify** [comment]
@kaushiktd - I sent an invite via Upwork. Please accept and I can complete the payment process. Thanks!

**Comment by kaushiktd** [comment]
Thank you @alexpensify 

I accepted the offer.

**Comment by alexpensify** [comment]
Done - I've completed the payment process in Upwork 

https://github.com/Expensify/App/issues/41763#issuecomment-2248625040 

@eVoloshchak - please submit a request in Chat, thanks!

**Comment by isabelastisser** [comment]
Thanks for handling this, @alexpensify!

**Comment by alexpensify** [comment]
You're welcome! @isabelastisser there is no more action to take here. 

@eVoloshchak - please submit a request in Chat. Thanks!

**Comment by JmillsExpensify** [comment]
$250 approved for @eVoloshchak

**Comment by QichenZhu** [comment]
Not sure this was actually fixed when it was closed. We ran into it again [here](https://github.com/Expensify/App/issues/86570) and fixed it [here](https://github.com/Expensify/App/pull/87834).
